[FX] Make output a non-special Node

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45775 [FX] Track use nodes in Node
* #45227 [FX] Mutable Graph APIs
* **#45599 [FX] Make output a non-special Node**

This will normalize how function outputs are handled to match out function inputs (i.e. outputs are now represented as `output` nodes, analogous to how inputs are handled as `placeholder` nodes). This makes things more uniform in several places. `Graph.output()` now inserts an `output` node into the Graph, rather than setting a special attribute.

Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

Commit:
[FX] Make output a non-special Node

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -80,6 +80,8 @@ def graph_copy(self, g : 'Graph'):
         val_map : Dict[Node, Node] = {}
         for node in g._nodes:
+            if node.op == 'output':
+                continue
```
`graph_copy` is a bit unfinished: it probably should take val_map as argument, and then return the output node of the graph being copied. If that were so this would turn into a return.

zdevito commented on torch/fx/graph.py:
```diff
@@ -157,8 +159,20 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
         self._mark_uses(result)
+        return self.create_node(op='output', target='output', args=(result,))
+
+    @property
```
I feel like this function should not exist in the version where `output` is not special. Especially as a property it suggests it is fast, but is actually linear in the size of the graph.

jamesr66a replied:
I agree with that, but one thing this is hiding is the fact that the output value is actually `args[0]` in the `output` node. I don't think that's nice and people will probably mess that up a lot if they have to manually extract the output value

zdevito commented on torch/fx/graph.py:
```diff
@@ -236,10 +251,18 @@ def python_code(self, root_module: str) -> Tuple[str, str, List[str]]:
                 assert isinstance(node.target, str)
                 body.append(f'{node.name} = {_format_target(root_module, node.target)}\n')
                 continue
+            elif node.op == 'output':
```
The separation of python code from the function has gotten weaker over time. We should consider just having python code generating the whole function, and that would allow us to just generate 'return out' as a string here without the special handling.

zdevito commented on torch/fx/proxy.py:
```diff
@@ -10,6 +10,9 @@
 class TracerBase:
     graph: Graph
 
+    def output(self, result : Argument) -> Node:
```
It was intentional that there only be a `create_node` function in the tracer. So this function should probably not exist.

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -149,7 +149,7 @@ def module_call_wrapper(mod, *args, **kwargs):
         try:
             torch.nn.Module.__call__ = module_call_wrapper
-            self.graph.output(self.create_arg(fn(*args)))
+            self.output(self.create_arg(fn(*args)))
```
This should call tracers create_node instead similar to how other things in symbolic trace work.

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

zdevito commented on test/fx/quantization.py:
```diff
@@ -240,7 +246,9 @@ def load_arg(a):
                 self.quants[node.name].observe(node, env)
 
-        return load_arg(self.graph.result)
+        if output_node is None:
```
Technically this could just go into the `if op == 'output'` statement, and cause the entire loop to terminate and return. It skips the check that there is only one return node, but we should be enforcing that anyway.

zdevito commented on torch/fx/graph.py:
```diff
@@ -236,10 +242,22 @@ def python_code(self, root_module: str) -> Tuple[str, str, List[str]]:
                 body.append(f'{node.name} = {_format_target(root_module, node.target)}\n')
                 continue
+            elif node.op == 'output':
+                if result_value:
```
It'd be conceptually simpler to just have output insert a `return x` into the body and delete the return formatting below. I wouldn't worry about enforcing single return semantics here, that is what the lint pass should be doing.

zdevito commented on torch/quantization/fx/quantize.py:
```diff
@@ -485,7 +493,9 @@ def is_observed(input_arg):
-        observed_graph.output(load_arg(model.graph.result))
+        if output_node is None:
+            raise RuntimeError('Graph had no output node!')
+        observed_graph.output(load_arg(output_node.args[0]))
```
Why not just put this in the loop?

jamesr66a replied:
The output value is actually also used at line 502. Not sure why this seemed to pass when I ran this in FBCode...

zdevito commented on torch/quantization/fx/quantize.py:
```diff
@@ -718,27 +734,38 @@ def is_quantized(node):
+            graph_output = map_arg(output_node.args[0], load_x)
         else:
-            graph_output = map_arg(model.graph.result, load_non_quantized)
+            graph_output = map_arg(output_node.args[0], load_non_quantized)
```
Again I think this can just go into the loop, right?

zdevito commented on torch/quantization/fx/quantize.py:
```diff
@@ -796,7 +829,9 @@ def load_arg(a):
-        folded_graph.output(load_arg(quantized_graph.result))
+        if output_node is None:
+            raise RuntimeError('Graph had no output node!')
+        folded_graph.output(load_arg(output_node.args[0]))
```
same here

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -74,13 +74,20 @@ def __init__(self):
         for node in g._nodes:
+            if node.op == 'output':
+                rv = map_arg(node.args[0], lambda n: val_map[n])
+                assert isinstance(rv, Node)
```
This assert doesn't have to be true. A graph can return a tuple.

zdevito replied:
Return type should be `Optional[Argument]` which is just `Argument` since it already contains None.

zdevito commented on torch/quantization/fx/quantize.py:
```diff
@@ -781,6 +791,9 @@ def load_arg(a):
         quantized_root = quantized
         quantized_graph = quantized.graph
         for node in quantized_graph.nodes:
+            if node.op == 'output':
```
This is probably not needed because there is a `node_copy` in the else branch which will do the same thing.

zdevito reviewed (approved):
Cool -- it looks like making output a node simplifies things compared to making parameters not nodes. So I think we should go with this approach. It should also simplified the mutation logic since you can always insert after a param or insert before a return.

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]

Commit:
Update on "[FX] Make output a non-special Node"


Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

[ghstack-poisoned]
